Fix auto-draft post permalink structure

youknowriad opened the pull request:
When loading the post editor the `permalink_template` property is not set properly on the post object (REST API) (it uses the plain post permalink instead of the pretty permalink). This causes the editor to not offer a way to edit the slug before actually saving the post which results in the permalink_template being set properly.

I tracked this down to `get_sample_permalink` and `get_permalink` return the plain permalink structure for auto-draft posts.

I'm not entirely certain about this fix as it's a bit low level, would appreciate a review from folks familiar with this.

Trac ticket: https://core.trac.wordpress.org/ticket/59283

youknowriad committed:
Fix auto-draft post permalink structure

youknowriad commented on src/wp-admin/includes/post.php:
```diff
@@ -1440,7 +1440,7 @@ function get_sample_permalink( $post, $title = null, $name = null ) {
 
 	// Hack: get_permalink() would return plain permalink for drafts, so we will fake that our post is published.
-	if ( in_array( $post->post_status, array( 'draft', 'pending', 'future' ), true ) ) {
+	if ( in_array( $post->post_status, array( 'auto-draft', 'draft', 'pending', 'future' ), true ) ) {
```
There was already a hack in place that switches the "post_status" to "publish" for "drafts" to ensure the permalink returned in the correct one by the underlying `get_permalink` function.

That said, adding `auto-draft` here was not sufficient because `get_permalink` would call `get_post_status` which itself calls `get_post` and this function would reset our "hacked post type" to the real post type aka `auto-draft`. So the second part of the fix is to prevent `get_post_status` from calling `get_post` if we already provide it with a post object.

I'm not entirely certain about all the consequences here and also I'm not sure why the hack works without this last change for `drafts` but not for `auto-drafts`

anton-vlasenko replied:
LGTM.

azaozz replied:
> the second part of the fix is to prevent get_post_status from calling get_post
> ...
> I'm not entirely certain about all the consequences here

Yep, thinking this is not such a good idea. It might result is some "unexplained" regressions in other unexpected code/places as (as far as I see) that deals with post caching.

Permalinks for non-published posts are a tricky thing. They are considered "temporary" as they may become unavailable by the time the post is published. On top of that the users can edit them. Generally auto-draft posts should never have a permalink as these posts "don't exist" in the real sense. They are simply placeholders indicating a possible intention to maybe have a post with that ID :)

Perhaps another option here is to ensure the post is saved as draft before trying to get or set a permalink for it?

youknowriad replied:
> Perhaps another option here is to ensure the post is saved as draft before trying to get or set a permalink for it?

I think that probably has bigger downsides as the user may leave the editor right away before saving or typing or anything and over time, we'll just end up with a huge numbers of draft posts. I believe that's why auto-drafts exist in the first place.

azaozz replied:
> the user may leave the editor right away before saving or typing or anything and over time, we'll just end up with a huge numbers of draft posts

Yes, it's possible. However if we have to set a permalink that means the user at least has typed a post title, i.e. they have started a new post (otherwise how would WP know what to use to make a permalink). Thinking that having post title is enough of a reason to save a draft. 

If a lot of these posts accumulate over time, they will all show in the Posts list table anyways, and the users will be able to either continue them or delete them.

youknowriad replied:
And why do we have a different behavior for pages? For pages, it's possible to set the slugs right away and this functions returns the right permalink right away. This is something I was not able to track down.

azaozz replied:
> For pages, it's possible to set the slugs right away and this functions returns the right permalink right away.

Right, looking at this again best imho would be to save as a draft if it is an auto-draft at that point, then do that "permalink hack". I.e. the change here will be that a new draft post is created when someone edits the permalink before doing anything else like adding title, content, etc.

Just adding `'auto-draft'` to that "hack" (i.e. allowing auto-drafts to have proper permalinks) may open some edge cases like permalinks for non-existing post IDs as old auto-drafts are deleted without warning (garbage collection).

youknowriad replied:
> Right, looking at this again best imho would be to save as a draft if it is an auto-draft at that point, then do that "permalink hack". I.e. the change here will be that a new draft post is created when someone edits the permalink before doing anything else like adding title, content, etc.

Unfortunately, that's not possible. `permalink_template` in the REST API response is what indicates to the site editor that the current post type and site actually supports pretty permalink, so if this property is not defined the UI to show the permalink input is not shown at all. Is there any other way (using the REST API) to check whether the current post supports pretty permalinks?

mcsf replied:
I took the liberty of pushing changes to `get_post_status` that should ensure that normalisation is only skipped when intended, by looking at the `WP_Post`'s `filter` instance variable. This follows the precedent set in `get_permalink` and `wp_force_plain_post_permalink`.

Does this assuage others' concerns too, or should other measures be taken?

azaozz replied:
> pushing changes to get_post_status that should ensure that normalisation is only skipped when intended

@mcsf Yea, looks like this removes most of the previous concerns about caching/refreshing the caches. As far as I see normalization can be skipped only for auto-drafts. This _shouldn't_ affect anything since auto-drafts are by definition "empty" or "pre-determined content" placeholders. On the other hand looking at the sheer number of (plugins) code that "does something" with WP_Post, there might be unexpected edge cases even with these restrictions, so perhaps some inline docs there would be useful.

> permalink_template in the REST API response is what indicates to the site editor that the current post type and site actually supports pretty permalink

@youknowriad Right. If that's the main problem here, why not fix it in the REST API instead of changing the `get_sample_permalink()` function? Maybe even there can be another way in the REST response to set or make the permalink template (if that's really needed) or just set it to `true`? Perhaps that would make it easier to control the UI from PHP. I realize this is a different direction after all this time, but do you think it may work better and eventually be useful in the future?

Another possibility may be to add the specific case of getting a permalink of an auto-draft to the `get_sample_permalink()` function. The simplest way would be to add an optional param to it. This would maintain full backwards compatibility with all the existing code that expects to not have sample permalinks on auto-drafts, and allow for this specific case to work?

youknowriad replied:
I think there are two questions in one here:

 1- I think the permalink_template value is broken in the REST API for auto-drafts. It's inconsistent with other post types for no valid reason and this is how it's computed https://github.com/WordPress/wordpress-develop/blob/4304be07f3042184dfb5c0390b83a4d05a01a43a/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1991-L1995 A call to `get_sample_permalink`. For me it's logical to fix it in `get_sample_permalink` but it's actually the purpose of this function. Now, if you think there's another way to solve this in the REST API, I'm all for it. REST API for me are just proxies to WP functions. 
 
 2- I do think post types endpoints lack a lot of "supports" keys. We face this a lot in Gutenberg and we end up often having to hard code post types. Things like:

 - Whether the post type support permalinks.
 - Whether the post type support auto-saves.

But there are a lot more and It would be great to add these things. It seems like a big effort.

azaozz replied:
> For me it's logical to fix it in get_sample_permalink()...

Right, makes sense. 

> if you think there's another way to solve this in the REST API

Was thinking the REST API can return just enough data (hard-coded?) to trigger the proper UI when the post is an auto-draft. However that will not run the `get_sample_permalink` filter. Seems running it for auto-drafts may be desirable. Returning minimal data from `get_sample_permalink()` would work too, also in cases outside the REST API.

May be overthinking this but as mentioned above for maximum backwards compatibility may be good to add another param to get_sample_permalink(). Perhaps something like:

`get_sample_permalink( $post, $title = null, $name = null, $for_auto_draft = false )`

Then use that in WP_REST_Posts_Controller:

`$sample_permalink = get_sample_permalink( $post->ID, $post->post_title, '', true );`

azaozz replied:
> post types endpoints lack a lot of "supports" keys. We face this a lot in Gutenberg

Thinking I've seen a discussion about outputting all the `post_type_supports` settings for a particular post type? At first look this doesn't seem hard to do, just "translating" the relevant part of the `$_wp_post_type_features` PHP global and adding it to a REST response. Guessing this has been tried before and there were problems? I'll be glad to help do that or resolve any problems if I can :)

youknowriad replied:
> May be overthinking this but as mentioned above for maximum backwards compatibility may be good to add another param to get_sample_permalink(). Perhaps something like:

I'm happy to add this, I don't mind it but I'm just noting that this comment already exists in the function. 

> 	// Hack: get_permalink() would return plain permalink for drafts, so we will fake that our post is published.

In other words, we don't have a `$for_draft` argument (or `$for_pending` in these cases). In other words, I wonder if "auto-drafts" were just forgotten in the first place in that "hack".

mcsf replied:
I think we're overthinking a bit. We should obviously be careful about back-compat and with areas of undefined behaviour that could affect users' edge cases out there... but IMO introducing `$for_draft` is going a bit far. Like Riad, I'm more inclined to think that auto-drafts were forgotten, and that fixing that shouldn't really affect anyone.

azaozz replied:
> think that auto-drafts were forgotten

Not exactly. The logic/design intent for auto-drafts is that they are not considered "posts" as the users have not provided any content for them yet. They are placeholders for future posts (that WP eventually auto-deletes) just to get the self-incrementing table row ID (the post_ID) from the DB. 

However this doesn't (shouldn't) affect the decision here. The only thing that matters is to keep back-compat as best as possible. Adding `auto-draft` to the post statuses that support sample permalinks changes the current behavior of `get_sample_permalink()`. I agree the change is very very small, but it's not impossible to cause some sort of issue somewhere. 

Generally the easiest thing to do in such cases is to make this a new "special case" and add some way to detect it in `get_sample_permalink()`. Adding another param seems to fit best? 

So I kind of prefer if there is another param despite that the chance for a back-compat problem is super small. (Perhaps this can also be seen as an example of how to handle similar cases in the future.)

azaozz replied:
> just noting that this comment already exists in the function...

Yea, I agree, not the best way to do this, by far (pretend that the post is published while calculating a permalink). I'm almost certain that this was done in order to keep the back-compat with some existing code/use case at the time :) 

This has been in core for so many years, all the quirks this code may be causing are now a part of how it works, and core and plugins have learned how to deal with them and are expecting them. 

This is, imho, one of the hardest things to account for when working with an old codebase :) 

> I wonder if "auto-drafts" were just forgotten in the first place in that "hack"

Frankly I think that auto-drafts didn't exist yet at the time this code was added.

youknowriad commented:
I'm happy to add tests and all later, just want to check whether this is the right fix or if there's something else that we should be doing.

anton-vlasenko commented on src/wp-includes/post.php:
```diff
@@ -1143,7 +1143,9 @@ function get_post_mime_type( $post = null ) {
  */
 function get_post_status( $post = null ) {
-	$post = get_post( $post );
+	if ( ! is_object( $post ) ) {
```
In my opinion, this change is a step in the right direction.
It doesn't make sense to retrieve an instance of `WP_Object` if it's already loaded. This also brings some performance improvement.
After examining the GitHub log, it appears there's no particular reason to call
```php
$post = get_post( $post );
```
every time the `get_post_status()` function is called.

azaozz replied:
Hmm, not so sure as `$post = get_post( $post );` deals with caching (all different kinds of it)? Seems removing it might bring some hard to explain/find regressions.

anton-vlasenko replied:
>  Seems removing it might bring some hard to explain/find regressions.

That was also my concern, @azaozz.
Some code might assume that `get_post($post)` is called even when there's objectively no need to (because `$post` is already an object).
However, I see the change that @youknowriad proposed as an improvement, but I also agree that it **_might_** lead to some hard to find regressions.

azaozz replied:
> objectively no need to (because $post is already an object).

Right, it may be an object but it may not be synced with the cache? Like cache may need to be updated, or (external) cache has changes in the meantime, or some external process/DB connection runs slow at the moment and this might be affected for some reason, or... who knows what else :)

peterwilsoncc replied:
`$post = get_post( $post )` is a very common pattern in WP-dev to normalize the post object. A regex search for `\$post\s*= get_post\( \$post \)` returns 121 results in `src` and probably misses a few minor variations.

I know it normalizes both integers and `WP_Post` objects but it may help with arrays too (Ozz, you probably have more intimate knowledge here).

If it's a pattern that could be improved, then it's way outside the scope of this PR.

anton-vlasenko reviewed (commented):
This change LGTM.
However, as with any code change, it's essential to test the modified function before deploying.

ofmarconi commented:
Is this not ready yet? Damn

mcsf committed:
get_post_status: respect 'sample' filter

Following the precedent established in related functions (e.g.
`get_permalink`, `wp_force_plain_post_permalink`), let `get_post_status`
skip post normalization when its argument is a WP_Post correctly stating
its intent with the `$post->filter === 'sample'` flag.

mcsf committed:
(remove unused $sample variable)

youknowriad commented on src/wp-includes/post.php:
```diff
@@ -1143,7 +1143,11 @@ function get_post_mime_type( $post = null ) {
 function get_post_status( $post = null ) {
-	$post = get_post( $post );
+	if ( is_object( $post ) && isset( $post->filter ) && 'sample' === $post->filter ) {
+		// Skip normalization
```
@mcsf Is there a reason for the empty "if". What if we just invert the condition?

mcsf replied:
> Is there a reason for the empty "if". What if we just invert the condition?

I thought it might be easier to read (and more closely follow the other functions that look at `filter`), but of course we can just flip it. No strong opinion here, I just went on an instinct.

youknowriad replied:
👌 I don't mind either.

azaozz replied:
Just a small suggestion (untested): If this code is staying (see above comments) perhaps can do `$post instanceof WP_Post` instead of `is_object( $post )` (seems a bit better/more precise?) and maybe add `'auto-draft' === $post->post_status` to the conditional?

youknowriad replied:
I can only apply the first part of the suggestion. The post status is forced to "publish" for these calls.

youknowriad commented:
This is looking very good to me @mcsf specially since this is a common pattern used in other places.

mcsf reviewed (approved):
Anyone feel free to disagree, but I'm adding my 👍

youknowriad committed:
Add unit test

youknowriad committed:
Apply suggestion

youknowriad commented:
I've added a test here. I think this is good to ship. I'll give it a try and happy to address follow-ups.

youknowriad commented:
commit https://core.trac.wordpress.org/changeset/58174
